Add resplitting functionality to Flower Datasets

adam-narozniak opened the pull request:
## Issue

The datasets downloaded from Hugging Face come with certain splits. Yet, users might want to use different divisions of the whole dataset, which is currently not possible.

### Description

``` python
fds = FederatedDataset(dataset="mnist", partitioners={"train": 10})
# then I can
fds.load_partition(10, "train")
```
But what if the dataset had three splits "train", "valid", "test" (not 2 like "mnist"). 
In that case, you might want to have a single dataset from which 10 partitions are created.

And the reverse might hold true for just 2-split dataset. What if a user want to have 3 splits. This is also impossible.

## Proposal
Enable the missing functionality described above (and add tests to make sure they are met).
* Introduce the `resplitter` keyword in the `FederatedDataset`
* Enable two major ways of creation of the new dataset (with different splits) 
* The first one `Callable[[DatasetDict], DatasetDict]]` You might perform as sophisticated change as you wish - just provide this as `resplitter` to the `FederatedDataset` (Note: all the checks to use that correctly are on the user side)
* The second option is `Dict[Tuple[str, ...], str]` (called for convenience `ResplitStrategy`). Here is an example `{("train", "valid"): "bigger_train"}`. We create a "bigger_train" split from the "train" and "valid" splits. From this object a `Resplitter` (newly introduced class) is created. That is essentially `Callable[[DatasetDict], DatasetDict]]` with additional check if the splits are used correctly (you can use only the existing splits, and you cannot create a new dataset that has two splits with the same name)
This works as follows:
 Frist option
```python
  def resplit(dataset: DatasetDict) -> DatasetDict:
      return DatasetDict(
          {
              "bigger_train": concatenate_datasets(
                  [dataset["train"], dataset["valid]]
              ),
              "test": dataset["test"]
          }
      )

  fds = FederatedDataset(
      dataset=self.dataset_name, partitioners={"bigger_train": 100}, resplitter=resplit
  )
  bigger_train = fds.load_full("bigger_train")
  # Or just load a partition
  partition_from_bigger = fds.load_partition("bigger_train": 100)
```
The second option (that does the same thing) **ResplitterStrategy** specification => internally **Resplitter** creation
```python
fds = FederatedDataset(
            dataset=self.dataset_name, partitioners={"bigger_train": 100}, resplitter={("train", "valid"): "bigger_train"}
     )
# if I made a mistake using here 
# e.g. resplitter = {("train", "split-that-does- not-exist"): "bigger_train"}
# I'll get a meaningful error
# similarly if I did sth like that 
# resplitter = {("train",): "new", ("valid", ): "new"}
# there can't be two splits named "new" so I'll get an error
bigger_train = fds.load_full("bigger_train")
# Or just load a partition
partition_from_bigger = fds.load_partition("bigger_train": 100)
```

danieljanes commented on datasets/flwr_datasets/merge_splitter.py:
```diff
@@ -0,0 +1,75 @@
+from datasets import Dataset, DatasetDict
+
+
+class MergeSplitter:
```
```suggestion
class MergeResplitter:
```

danieljanes commented on datasets/flwr_datasets/federated_dataset.py:
```diff
@@ -35,6 +38,8 @@ class FederatedDataset:
     dataset: str
         The name of the dataset in the Hugging Face Hub.
+    resplitter: Optional[Union[Resplitter, Dict[Tuple[str, ...], str]]]
+        Resplit strategy or custom Callable that transforms the dataset.
```
```suggestion
        Resplit strategy or custom `Callable` that transforms splits in the `DatasetDict`.
```

danieljanes commented on datasets/flwr_datasets/federated_dataset.py:
```diff
@@ -153,3 +163,22 @@ def _assign_dataset_to_partitioner(self, split: str) -> None:
             self._partitioners[split].dataset = self._dataset[split]
+
+    def _resplit_dataset_if_needed(self) -> None:
+        # this can't be called many times
+        # either a new attribute is needed e.g. resplit_dataset
+        # or a bool flag that the resplit happened
```
```suggestion
        # The actual re-splitting can't be done more than once.
        # The attribute `_resplit` indicates that the resplit happened.
```

danieljanes commented on datasets/flwr_datasets/merge_splitter.py:
```diff
@@ -0,0 +1,75 @@
+
+
+class MergeSplitter:
+    """Create a new dataset splits according to the `resplit_strategy`.
+
+    The dataset comes with some predefined splits e.g. "train", "valid" and "test". This
+    class allows you to create a new dataset with splits created according to your needs
+    specified in `resplit_strategy`.
+
+    Parameters
+    ----------
+    resplit_strategy: ResplitStrategy
+        Dictionary with keys - tuples of the current split names to values - the desired
+        split names
+    """
```
Let's reword this to make it more obvious that this `Resplitter` implementation merges existing splits (and assigns a new name to the newly created split).

adam-narozniak replied:
Done

danieljanes commented on datasets/flwr_datasets/merge_splitter.py:
```diff
@@ -0,0 +1,75 @@
+
+    def __init__(
+        self,
+        resplit_strategy: Dict[Tuple[str, ...], str],
```
Should we call this `merge_config` or `merge_strategy` or something similar? It's perhaps best to avoid `strategy` since this term is already used for a different concept in Flower.

adam-narozniak replied:
`merge_config` sounds good

adam-narozniak commented:
Maybe the `MergeResplitter` should be simply `Merger` or `SplitMerger`? WDYT?

jafermarq commented on datasets/flwr_datasets/merge_resplitter.py:
```diff
@@ -0,0 +1,91 @@
+    def _check_duplicate_desired_splits(self) -> None:
+        """Check for duplicate desired split names."""
+        desired_splits = list(self._merge_config.values())
+        duplicates = [
```
Should we also add checks (or throw a warning) if the original splits are present in more than one of the new ones ? For instance the below will run without throwing error:
```
splitter = MergeResplitter(merge_config={('train','test'): 'fused', ('test',): 'test'})
```

adam-narozniak replied:
Good catch. I'm not sure. 
Flexibility: I'd go for a Warning. I think that the most cases it would be a problem, but maybe there are also intended usages of that? I am not sure. 
Usage-wise: That's probably the utility for beginners, so this will likely be a bug -> Error

danieljanes replied:
I like the idea of a warning, but still allowing the user to do it.

jafermarq replied:
Yeah, maybe warning is the reasonable option here. I agree with Daniel's view of not restricting users with this.

jafermarq reviewed (commented):
Very cool functionality. I just left a single comment.

jafermarq commented:
> Also, one more thing. I don't think the `Dict[Tuple[str, ...], str]` (the `Tuple` as the key) is that common. It matches more naturally the flow of operation: FROM (some keywords that already exist) -----create---> TO (new keywords). WDYT? We might change that.

This felt fine when I wast testing the splitter functionality.

danieljanes commented on datasets/flwr_datasets/federated_dataset.py:
```diff
@@ -35,6 +38,8 @@ class FederatedDataset:
     dataset: str
         The name of the dataset in the Hugging Face Hub.
+    resplitter: Optional[Union[Resplitter, Dict[Tuple[str, ...], str]]]
+        Resplit strategy or `Callable` that transforms splits in the `DatasetDict`.
```
```suggestion
        `Callable` that transforms `DatasetDict` splits, or configuration dict for `MergeResplitter`.
```

danieljanes commented on datasets/flwr_datasets/federated_dataset.py:
```diff
@@ -153,3 +163,21 @@ def _assign_dataset_to_partitioner(self, split: str) -> None:
+        if self._dataset is None:
+            raise ValueError("The dataset resplit should happen after the download.")
+        if self._resplitter:
+            resplitter: Resplitter
+            if isinstance(self._resplitter, Dict):
+                resplitter = MergeResplitter(merge_config=self._resplitter)
+            else:
+                resplitter = self._resplitter
```
```suggestion
            if isinstance(self._resplitter, Dict):
                self._resplitter = MergeResplitter(merge_config=self._resplitter)
            resplitter: Resplitter = self._resplitter
```

danieljanes commented on datasets/flwr_datasets/merge_resplitter.py:
```diff
@@ -0,0 +1,91 @@
+"""MergeResplitter class for Flower Datasets."""
+import collections
+from typing import Dict, List, Tuple
```
The copyright notice is missing in this file.

```suggestion
"""MergeResplitter class for Flower Datasets."""


import collections
from typing import Dict, List, Tuple
```

danieljanes commented on datasets/flwr_datasets/merge_resplitter_test.py:
```diff
@@ -0,0 +1,129 @@
+"""Resplitter tests."""
+import unittest
+from typing import Dict, Tuple
```
Copyright notices is missing

```suggestion
"""Resplitter tests."""


import unittest
from typing import Dict, Tuple
```

adam-narozniak commented:
I've switched the keys and values. I also fixed the resplit_dataset_if_needed (I violated the single responsibility principle there and it had complex initialization and then resplit functionality, inti now moved to utils; that also created circular import of Resplitter so I created a common.typing which of which the name can be changed but typing alone didn't work so if common.typing is not ok, then maybe types.py?)

danieljanes commented on datasets/flwr_datasets/common/typing.py:
```diff
@@ -0,0 +1,20 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Flower Datasets type definitions."""
+from typing import Callable
```
```suggestion
"""Flower Datasets type definitions."""


from typing import Callable
```
